### DISCUSSION TITLE
How can I make exceptions to the whitespace rule?

### DISCUSSION THREAD

**Discussion by Thesecondbestname** [pest-parser/pest#769] [Q&A] [answered]
I want my parser to ignore every whitespace and newline except between two expressions or after a use statement. I'm sure I can work the two subsequent expressions out but how do I make exceptions to the whitespace rule? I've thought about using the stack here to as in adding a `PUSH("USE") ~ POP` at the end of the use statement and check the stack inside the whitespace rule but I don't know how to check the stack for a specific value. I am very new so please be gentle :face_with_head_bandage:

**Comment by tomtau** [comment, marked as answer]
I guess defining rules as atomic and non-atomic can help: http://pest.rs/book/grammars/syntax.html#implicit-whitespace ?

**Reply by Thesecondbestname** [reply]
Is this the only way? Because I think I'd have to implement the whitespace rule manually where I just want to require one...

**Reply by Thesecondbestname** [reply]
I think it could be done via the stack but I don't know how to match on `PEEK`. Is that possible? Something like `PEEK == "USE"` ?

**Reply by tomtau** [reply]
The stack manipulation is for matching the *same text*: https://pest.rs/book/grammars/syntax.html#the-stack-wip

so `PUSH("USE")` and then `POP` or `PEEK` means to match: `USE ... USE`

**Reply by Thesecondbestname** [reply]
Thanks a lot! Thanks to a critical error in my thought process it took me a day to understand how to do it using atomic rules.
